e2e/fixtures.js

```diff
@@ -16,7 +23,32 @@ function consoleMessageToString(msg) {
     at (${url} ${lineNumber}:${columnNumber})`;
 }
 
+//The following is based on https://github.com/mxschmitt/playwright-test-coverage
+const istanbulCLIOutput = path.join(process.cwd(), '.nyc_output');
+
+function generateUUID() {
+    return crypto.randomBytes(16).toString('hex');
+}
+
 exports.test = base.test.extend({
+    //The following is based on https://github.com/mxschmitt/playwright-test-coverage
+    context: async ({ context }, use) => {
+        await context.addInitScript(() =>
+            window.addEventListener('beforeunload', () =>
+                (window).collectIstanbulCoverage(JSON.stringify((window).__coverage__))
+            )
+        );
+        await fs.promises.mkdir(istanbulCLIOutput, { recursive: true });
+        await context.exposeFunction('collectIstanbulCoverage', (coverageJSON) => {
+            if (coverageJSON) {
+                fs.writeFileSync(path.join(istanbulCLIOutput, `playwright_coverage_${generateUUID()}.json`), coverageJSON);
```

Comment:
i think you can just do `await fs.promises.writeFile(...` and that way not use a synchronous block here